test/ClientWidgetApi-test.ts

```diff
@@ -0,0 +1,348 @@
+
+afterEach(() => {
+    jest.resetAllMocks();
+})
+
+function createRoomEvent(event: Partial<IRoomEvent> = {}): IRoomEvent {
+    return {
+        type: 'm.room.message',
+        sender: 'user-id',
+        content: {},
+        origin_server_ts: 0,
+        event_id: 'id-0',
+        room_id: '!room-id',
+        unsigned: {},
+        ...event,
+    };
+}
+
+describe('ClientWidgetApi', () => {
+    let capabilities: Capability[];
+    let iframe: HTMLIFrameElement;
+    let driver: jest.Mocked<WidgetDriver>;
+    let clientWidgetApi: ClientWidgetApi;
+    let transport: PostmessageTransport;
+    let emitEvent: Parameters<PostmessageTransport["on"]>["1"];
+
+    async function loadIframe(caps: Capability[] = []) {
+        capabilities = caps;
+
+        const ready = new Promise<void>(resolve => {
+            clientWidgetApi.once('ready', resolve);
+        });
+
+        iframe.dispatchEvent(new Event('load'));
+
+        await ready;
+    }
+
+    beforeEach(() => {
+        capabilities = [];
+        iframe = document.createElement('iframe');
+        document.body.appendChild(iframe);
+
+        driver = {
+            readEventRelations: jest.fn(),
+            validateCapabilities: jest.fn(),
+        } as Partial<WidgetDriver> as jest.Mocked<WidgetDriver>;
+
+        clientWidgetApi = new ClientWidgetApi(
+            new Widget({
+                id: "test",
+                creatorUserId: "@alice:example.org",
+                type: "example",
+                url: "https://example.org",
+            }),
+            iframe,
+            driver,
+        );
+
+        ([transport] = jest.mocked(PostmessageTransport).mock.instances);
+        emitEvent = jest.mocked(transport.on).mock.calls[0][1];
+
+        jest.mocked(transport.send).mockResolvedValue({});
+        jest.mocked(driver.validateCapabilities).mockImplementation(
+            async () => new Set(capabilities),
+        );
+    });
+
+    afterEach(() => {
+        clientWidgetApi.stop();
+        iframe.remove();
+    });
+
+    it('should initiate capabilities', async () => {
+        await loadIframe(['m.always_on_screen']);
+
+        expect(clientWidgetApi.hasCapability('m.always_on_screen')).toBe(true);
+        expect(clientWidgetApi.hasCapability('m.sticker')).toBe(false);
+    });
+
+    describe('org.matrix.msc3869.read_relations action', () => {
+        it('should present as supported api version', () => {
+            const event: ISupportedVersionsActionRequest = {
+                api: WidgetApiDirection.FromWidget,
+                widgetId: 'test',
+                requestId: '0',
+                action: WidgetApiFromWidgetAction.SupportedApiVersions,
+                data: {},
+            };
+
+            emitEvent(new CustomEvent('', { detail: event }));
+
+            expect(transport.reply).toBeCalledWith(event, {
+                supported_versions: expect.arrayContaining([
+                    'org.matrix.msc3869',
```

Comment:
I would recommend using the `UnstableApiVersion` enum value instead

Reply:
I usually prefer to use strings over enums in tests to catch unintended breaking changes on changes to the enum values. Like a kind of low-effort api contract testing. But I'm fine to match with the project style 👍.